// websocket13/dist/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.StreamedOutgoingMessage = exports.StreamedIncomingMessage = exports.WebSocketServerConnection = exports.WebSocketServer = exports.WebSocket = exports.StatusCode = exports.FrameType = exports.State = void 0;
const State_1 = __importDefault(require("./enums/State"));
exports.State = State_1.default;
const FrameType_1 = __importDefault(require("./enums/FrameType"));
exports.FrameType = FrameType_1.default;
const StatusCode_1 = __importDefault(require("./enums/StatusCode"));
exports.StatusCode = StatusCode_1.default;
const WebSocket_1 = __importDefault(require("./WebSocket"));
exports.WebSocket = WebSocket_1.default;
const WebSocketServer_1 = __importDefault(require("./WebSocketServer"));
exports.WebSocketServer = WebSocketServer_1.default;
const WebSocketServerConnection_1 = __importDefault(require("./WebSocketServerConnection"));
exports.WebSocketServerConnection = WebSocketServerConnection_1.default;
const StreamedIncomingMessage_1 = __importDefault(require("./streams/StreamedIncomingMessage"));
exports.StreamedIncomingMessage = StreamedIncomingMessage_1.default;
const StreamedOutgoingMessage_1 = __importDefault(require("./streams/StreamedOutgoingMessage"));
exports.StreamedOutgoingMessage = StreamedOutgoingMessage_1.default;
// Also provide a default export
exports.default = {
    // Enums
    State: State_1.default,
    FrameType: FrameType_1.default,
    StatusCode: StatusCode_1.default,
    // Classes that consumers directly instantiate
    WebSocket: WebSocket_1.default,
    WebSocketServer: WebSocketServer_1.default
};
