// websocket13/dist/streams/StreamedOutgoingMessage.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const bytebuffer_1 = __importDefault(require("bytebuffer"));
const stream_1 = require("stream");
const FrameType_1 = __importDefault(require("../enums/FrameType"));
class StreamedOutgoingMessage extends stream_1.Writable {
    constructor(socket, type) {
        super();
        this.started = false;
        this.finished = false;
        this.dataSent = false;
        this.type = type;
        this._socket = socket;
        this._chunks = [];
        this.on('finish', () => {
            this.finished = true;
            this._emptyQueue();
        });
        // @ts-ignore
        this.setDefaultEncoding('binary');
    }
    _start() {
        if (this.started) {
            return;
        }
        this.started = true;
        for (let i = 0; i < this._chunks.length; i++) {
            if (this._chunks[i].callback) {
                this._chunks[i].callback(null);
                this._chunks[i].callback = null;
            }
        }
        this._emptyQueue();
    }
    _emptyQueue() {
        if (!this.started || this._chunks.length == 0) {
            return;
        }
        let chunks = this._chunks.slice(0, this.finished ? this._chunks.length : this._chunks.length - 1);
        if (chunks.length == 0) {
            return;
        }
        // No reason not to combine all those chunks into one
        let buffer = Buffer.concat(chunks.map(chunk => chunk.payload));
        this._socket._sendFrame({
            FIN: !!this.finished,
            RSV1: false,
            RSV2: false,
            RSV3: false,
            opcode: !this.dataSent ? this.type : FrameType_1.default.Continuation,
            payloadLength: buffer.length,
            payload: buffer
        }, true);
        this.dataSent = true;
        chunks.filter(chunk => !!chunk.callback).forEach(chunk => chunk.callback(null));
        this._chunks.splice(0, chunks.length);
        this._socket._processQueue();
    }
    _write(chunk, encoding, callback) {
        if (this.type == FrameType_1.default.Data.Binary && typeof chunk === 'string') {
            callback(new Error('Cannot send a string through a binary frame.'));
            this.end();
            return;
        }
        if (bytebuffer_1.default.isByteBuffer(chunk)) {
            chunk = chunk.toBuffer();
        }
        // We have to call the callback now or else remaining chunks don't make their way to us
        if (this.started) {
            callback(null);
        }
        this._chunks.push({
            payload: chunk,
            callback: !this.started ? callback : null
        });
        this._emptyQueue();
        this._socket._processQueue();
    }
}
exports.default = StreamedOutgoingMessage;
