// @doctormckay/stdlib/lib/data_structures/LeastUsedCache.js
"use strict";
var __classPrivateFieldSet = (this && this.__classPrivateFieldSet) || function (receiver, state, value, kind, f) {
    if (kind === "m") throw new TypeError("Private method is not writable");
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a setter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot write private member to an object whose class did not declare it");
    return (kind === "a" ? f.call(receiver, value) : f ? f.value = value : state.set(receiver, value)), value;
};
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
};
var _LeastUsedCache_entries, _LeastUsedCache_lastAccess, _LeastUsedCache_maxItems, _LeastUsedCache_gcInterval, _LeastUsedCache_lastGc;
Object.defineProperty(exports, "__esModule", { value: true });
class LeastUsedCache {
    /**
     * Construct a new LeastUsedCache.
     * @param {int} maxItems - Maximum number of items allowed in the cache before stuff will start being pruned
     * @param {int} gcInterval - Time in milliseconds between garbage collections
     * @constructor
     */
    constructor(maxItems, gcInterval) {
        _LeastUsedCache_entries.set(this, void 0);
        _LeastUsedCache_lastAccess.set(this, void 0);
        _LeastUsedCache_maxItems.set(this, void 0);
        _LeastUsedCache_gcInterval.set(this, void 0);
        _LeastUsedCache_lastGc.set(this, void 0);
        __classPrivateFieldSet(this, _LeastUsedCache_entries, {}, "f");
        __classPrivateFieldSet(this, _LeastUsedCache_lastAccess, {}, "f");
        __classPrivateFieldSet(this, _LeastUsedCache_maxItems, maxItems, "f");
        __classPrivateFieldSet(this, _LeastUsedCache_gcInterval, gcInterval, "f");
        __classPrivateFieldSet(this, _LeastUsedCache_lastGc, Date.now(), "f");
    }
    /**
     * Add an entry to the cache.
     * @param {string} key - The key under which this entry should be stored
     * @param {*} val - The value to store in this entry
     */
    add(key, val) {
        __classPrivateFieldGet(this, _LeastUsedCache_entries, "f")[key] = val;
        __classPrivateFieldGet(this, _LeastUsedCache_lastAccess, "f")[key] = Date.now();
        this.checkGC();
    }
    /**
     * Get the entry stored in the cache under a particular key.
     * @param {string} key - The key to retrieve
     * @return {null|*} value if present, null if not
     */
    get(key) {
        if (typeof __classPrivateFieldGet(this, _LeastUsedCache_entries, "f")[key] != 'undefined') {
            __classPrivateFieldGet(this, _LeastUsedCache_lastAccess, "f")[key] = Date.now();
            this.checkGC();
            return __classPrivateFieldGet(this, _LeastUsedCache_entries, "f")[key];
        }
        else {
            return null;
        }
    }
    /**
     * Delete an entry from the cache.
     * @param {string} key
     */
    delete(key) {
        delete __classPrivateFieldGet(this, _LeastUsedCache_entries, "f")[key];
        delete __classPrivateFieldGet(this, _LeastUsedCache_lastAccess, "f")[key];
        this.checkGC();
    }
    /**
     * Get a list of all keys in the cache.
     * @returns {string[]}
     */
    getKeys() {
        this.checkGC();
        return Object.keys(__classPrivateFieldGet(this, _LeastUsedCache_entries, "f"));
    }
    /**
     * Check if a garbage collection is necessary and if so, do it.
     */
    checkGC() {
        if (Date.now() - __classPrivateFieldGet(this, _LeastUsedCache_lastGc, "f") >= __classPrivateFieldGet(this, _LeastUsedCache_gcInterval, "f")) {
            this.gc();
        }
    }
    /**
     * Collect garbage and delete anything over the limit that hasn't been accessed in a while.
     */
    gc() {
        __classPrivateFieldSet(this, _LeastUsedCache_lastGc, Date.now(), "f");
        let keys = this.getKeys();
        if (keys.length <= __classPrivateFieldGet(this, _LeastUsedCache_maxItems, "f")) {
            return; // we aren't over the limit, so nothing to do
        }
        // sort the keys so that the least-frequently-accessed ones are at the end
        keys.sort((a, b) => __classPrivateFieldGet(this, _LeastUsedCache_lastAccess, "f")[a] > __classPrivateFieldGet(this, _LeastUsedCache_lastAccess, "f")[b] ? -1 : 1);
        keys.slice(keys.length - (keys.length - __classPrivateFieldGet(this, _LeastUsedCache_maxItems, "f"))).forEach((key) => {
            this.delete(key);
        });
    }
}
exports.default = LeastUsedCache;
_LeastUsedCache_entries = new WeakMap(), _LeastUsedCache_lastAccess = new WeakMap(), _LeastUsedCache_maxItems = new WeakMap(), _LeastUsedCache_gcInterval = new WeakMap(), _LeastUsedCache_lastGc = new WeakMap();
