// @doctormckay/stdlib/lib/data_structures/Stack.js
"use strict";
var __classPrivateFieldSet = (this && this.__classPrivateFieldSet) || function (receiver, state, value, kind, f) {
    if (kind === "m") throw new TypeError("Private method is not writable");
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a setter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot write private member to an object whose class did not declare it");
    return (kind === "a" ? f.call(receiver, value) : f ? f.value = value : state.set(receiver, value)), value;
};
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
};
var _Stack_length, _Stack_tail;
Object.defineProperty(exports, "__esModule", { value: true });
class Stack {
    /**
     * Create a new Stack. A Stack is a FILO data structure, in which all you can do is append items and remove items from
     * the back. Under the hood, this is implemented with a linked list (LL).
     * @constructor
     */
    constructor() {
        _Stack_length.set(this, void 0);
        _Stack_tail.set(this, void 0);
        __classPrivateFieldSet(this, _Stack_length, 0, "f");
        __classPrivateFieldSet(this, _Stack_tail, null, "f");
    }
    get length() {
        return __classPrivateFieldGet(this, _Stack_length, "f");
    }
    /**
     * Push a new item to the top of the stack.
     * @param {*} item - The item to push into the stack
     * @return {int} The new length of the stack
     */
    push(item) {
        var _a;
        // Create the LL node. The "prev" pointer is the existing tail of the list.
        __classPrivateFieldSet(this, _Stack_tail, { data: item, prev: __classPrivateFieldGet(this, _Stack_tail, "f") }, "f");
        return __classPrivateFieldSet(this, _Stack_length, (_a = __classPrivateFieldGet(this, _Stack_length, "f"), ++_a), "f");
    }
    /**
     * Remove the top element from the stack and return it.
     * @return {*} The top item in the stack. Null if the stack is empty.
     */
    pop() {
        var _a;
        if (!__classPrivateFieldGet(this, _Stack_tail, "f")) {
            return null;
        }
        let entry = __classPrivateFieldGet(this, _Stack_tail, "f").data;
        __classPrivateFieldSet(this, _Stack_tail, __classPrivateFieldGet(this, _Stack_tail, "f").prev, "f");
        __classPrivateFieldSet(this, _Stack_length, (_a = __classPrivateFieldGet(this, _Stack_length, "f"), _a--, _a), "f");
        return entry;
    }
    /**
     * Empty this stack by removing all items in it.
     */
    empty() {
        __classPrivateFieldSet(this, _Stack_tail, null, "f");
        __classPrivateFieldSet(this, _Stack_length, 0, "f");
    }
}
exports.default = Stack;
_Stack_length = new WeakMap(), _Stack_tail = new WeakMap();
