// steam-session/dist/enums-steam/EAuthTokenRevokeAction.js
"use strict";
/* eslint-disable */
// Auto-generated by generate-enums script on Fri Jun 30 2023 01:15:12 GMT-0400 (Eastern Daylight Time)
Object.defineProperty(exports, "__esModule", { value: true });
var EAuthTokenRevokeAction;
(function (EAuthTokenRevokeAction) {
    EAuthTokenRevokeAction[EAuthTokenRevokeAction["Logout"] = 0] = "Logout";
    EAuthTokenRevokeAction[EAuthTokenRevokeAction["Permanent"] = 1] = "Permanent";
    EAuthTokenRevokeAction[EAuthTokenRevokeAction["Replaced"] = 2] = "Replaced";
    EAuthTokenRevokeAction[EAuthTokenRevokeAction["Support"] = 3] = "Support";
    EAuthTokenRevokeAction[EAuthTokenRevokeAction["Consume"] = 4] = "Consume";
    EAuthTokenRevokeAction[EAuthTokenRevokeAction["NonRememberedLogout"] = 5] = "NonRememberedLogout";
    EAuthTokenRevokeAction[EAuthTokenRevokeAction["NonRememberedPermanent"] = 6] = "NonRememberedPermanent";
    EAuthTokenRevokeAction[EAuthTokenRevokeAction["Automatic"] = 7] = "Automatic";
})(EAuthTokenRevokeAction || (EAuthTokenRevokeAction = {}));
exports.default = EAuthTokenRevokeAction;
