// steam-session/dist/enums-steam/EResult.js
"use strict";
/* eslint-disable */
// Auto-generated by generate-enums script on Thu Nov 02 2023 00:05:05 GMT-0400 (Eastern Daylight Time)
Object.defineProperty(exports, "__esModule", { value: true });
/**
 * Contains possible result codes. This is a very large enum that is used throughout Steam, so most values in this enum won't be relevant when authenticating.
 */
var EResult;
(function (EResult) {
    EResult[EResult["Invalid"] = 0] = "Invalid";
    EResult[EResult["OK"] = 1] = "OK";
    EResult[EResult["Fail"] = 2] = "Fail";
    EResult[EResult["NoConnection"] = 3] = "NoConnection";
    EResult[EResult["InvalidPassword"] = 5] = "InvalidPassword";
    EResult[EResult["LoggedInElsewhere"] = 6] = "LoggedInElsewhere";
    EResult[EResult["InvalidProtocolVer"] = 7] = "InvalidProtocolVer";
    EResult[EResult["InvalidParam"] = 8] = "InvalidParam";
    EResult[EResult["FileNotFound"] = 9] = "FileNotFound";
    EResult[EResult["Busy"] = 10] = "Busy";
    EResult[EResult["InvalidState"] = 11] = "InvalidState";
    EResult[EResult["InvalidName"] = 12] = "InvalidName";
    EResult[EResult["InvalidEmail"] = 13] = "InvalidEmail";
    EResult[EResult["DuplicateName"] = 14] = "DuplicateName";
    EResult[EResult["AccessDenied"] = 15] = "AccessDenied";
    EResult[EResult["Timeout"] = 16] = "Timeout";
    EResult[EResult["Banned"] = 17] = "Banned";
    EResult[EResult["AccountNotFound"] = 18] = "AccountNotFound";
    EResult[EResult["InvalidSteamID"] = 19] = "InvalidSteamID";
    EResult[EResult["ServiceUnavailable"] = 20] = "ServiceUnavailable";
    EResult[EResult["NotLoggedOn"] = 21] = "NotLoggedOn";
    EResult[EResult["Pending"] = 22] = "Pending";
    EResult[EResult["EncryptionFailure"] = 23] = "EncryptionFailure";
    EResult[EResult["InsufficientPrivilege"] = 24] = "InsufficientPrivilege";
    EResult[EResult["LimitExceeded"] = 25] = "LimitExceeded";
    EResult[EResult["Revoked"] = 26] = "Revoked";
    EResult[EResult["Expired"] = 27] = "Expired";
    EResult[EResult["AlreadyRedeemed"] = 28] = "AlreadyRedeemed";
    EResult[EResult["DuplicateRequest"] = 29] = "DuplicateRequest";
    EResult[EResult["AlreadyOwned"] = 30] = "AlreadyOwned";
    EResult[EResult["IPNotFound"] = 31] = "IPNotFound";
    EResult[EResult["PersistFailed"] = 32] = "PersistFailed";
    EResult[EResult["LockingFailed"] = 33] = "LockingFailed";
    EResult[EResult["LogonSessionReplaced"] = 34] = "LogonSessionReplaced";
    EResult[EResult["ConnectFailed"] = 35] = "ConnectFailed";
    EResult[EResult["HandshakeFailed"] = 36] = "HandshakeFailed";
    EResult[EResult["IOFailure"] = 37] = "IOFailure";
    EResult[EResult["RemoteDisconnect"] = 38] = "RemoteDisconnect";
    EResult[EResult["ShoppingCartNotFound"] = 39] = "ShoppingCartNotFound";
    EResult[EResult["Blocked"] = 40] = "Blocked";
    EResult[EResult["Ignored"] = 41] = "Ignored";
    EResult[EResult["NoMatch"] = 42] = "NoMatch";
    EResult[EResult["AccountDisabled"] = 43] = "AccountDisabled";
    EResult[EResult["ServiceReadOnly"] = 44] = "ServiceReadOnly";
    EResult[EResult["AccountNotFeatured"] = 45] = "AccountNotFeatured";
    EResult[EResult["AdministratorOK"] = 46] = "AdministratorOK";
    EResult[EResult["ContentVersion"] = 47] = "ContentVersion";
    EResult[EResult["TryAnotherCM"] = 48] = "TryAnotherCM";
    EResult[EResult["PasswordRequiredToKickSession"] = 49] = "PasswordRequiredToKickSession";
    EResult[EResult["AlreadyLoggedInElsewhere"] = 50] = "AlreadyLoggedInElsewhere";
    EResult[EResult["Suspended"] = 51] = "Suspended";
    EResult[EResult["Cancelled"] = 52] = "Cancelled";
    EResult[EResult["DataCorruption"] = 53] = "DataCorruption";
    EResult[EResult["DiskFull"] = 54] = "DiskFull";
    EResult[EResult["RemoteCallFailed"] = 55] = "RemoteCallFailed";
    EResult[EResult["PasswordNotSet"] = 56] = "PasswordNotSet";
    EResult[EResult["PasswordUnset"] = 56] = "PasswordUnset";
    EResult[EResult["ExternalAccountUnlinked"] = 57] = "ExternalAccountUnlinked";
    EResult[EResult["PSNTicketInvalid"] = 58] = "PSNTicketInvalid";
    EResult[EResult["ExternalAccountAlreadyLinked"] = 59] = "ExternalAccountAlreadyLinked";
    EResult[EResult["RemoteFileConflict"] = 60] = "RemoteFileConflict";
    EResult[EResult["IllegalPassword"] = 61] = "IllegalPassword";
    EResult[EResult["SameAsPreviousValue"] = 62] = "SameAsPreviousValue";
    EResult[EResult["AccountLogonDenied"] = 63] = "AccountLogonDenied";
    EResult[EResult["CannotUseOldPassword"] = 64] = "CannotUseOldPassword";
    EResult[EResult["InvalidLoginAuthCode"] = 65] = "InvalidLoginAuthCode";
    EResult[EResult["AccountLogonDeniedNoMailSent"] = 66] = "AccountLogonDeniedNoMailSent";
    EResult[EResult["AccountLogonDeniedNoMail"] = 66] = "AccountLogonDeniedNoMail";
    EResult[EResult["HardwareNotCapableOfIPT"] = 67] = "HardwareNotCapableOfIPT";
    EResult[EResult["IPTInitError"] = 68] = "IPTInitError";
    EResult[EResult["ParentalControlRestricted"] = 69] = "ParentalControlRestricted";
    EResult[EResult["FacebookQueryError"] = 70] = "FacebookQueryError";
    EResult[EResult["ExpiredLoginAuthCode"] = 71] = "ExpiredLoginAuthCode";
    EResult[EResult["IPLoginRestrictionFailed"] = 72] = "IPLoginRestrictionFailed";
    EResult[EResult["AccountLocked"] = 73] = "AccountLocked";
    EResult[EResult["AccountLockedDown"] = 73] = "AccountLockedDown";
    EResult[EResult["AccountLogonDeniedVerifiedEmailRequired"] = 74] = "AccountLogonDeniedVerifiedEmailRequired";
    EResult[EResult["NoMatchingURL"] = 75] = "NoMatchingURL";
    EResult[EResult["BadResponse"] = 76] = "BadResponse";
    EResult[EResult["RequirePasswordReEntry"] = 77] = "RequirePasswordReEntry";
    EResult[EResult["ValueOutOfRange"] = 78] = "ValueOutOfRange";
    EResult[EResult["UnexpectedError"] = 79] = "UnexpectedError";
    EResult[EResult["Disabled"] = 80] = "Disabled";
    EResult[EResult["InvalidCEGSubmission"] = 81] = "InvalidCEGSubmission";
    EResult[EResult["RestrictedDevice"] = 82] = "RestrictedDevice";
    EResult[EResult["RegionLocked"] = 83] = "RegionLocked";
    EResult[EResult["RateLimitExceeded"] = 84] = "RateLimitExceeded";
    EResult[EResult["AccountLogonDeniedNeedTwoFactorCode"] = 85] = "AccountLogonDeniedNeedTwoFactorCode";
    EResult[EResult["AccountLoginDeniedNeedTwoFactor"] = 85] = "AccountLoginDeniedNeedTwoFactor";
    EResult[EResult["ItemOrEntryHasBeenDeleted"] = 86] = "ItemOrEntryHasBeenDeleted";
    EResult[EResult["ItemDeleted"] = 86] = "ItemDeleted";
    EResult[EResult["AccountLoginDeniedThrottle"] = 87] = "AccountLoginDeniedThrottle";
    EResult[EResult["TwoFactorCodeMismatch"] = 88] = "TwoFactorCodeMismatch";
    EResult[EResult["TwoFactorActivationCodeMismatch"] = 89] = "TwoFactorActivationCodeMismatch";
    EResult[EResult["AccountAssociatedToMultiplePlayers"] = 90] = "AccountAssociatedToMultiplePlayers";
    EResult[EResult["AccountAssociatedToMultiplePartners"] = 90] = "AccountAssociatedToMultiplePartners";
    EResult[EResult["NotModified"] = 91] = "NotModified";
    EResult[EResult["NoMobileDeviceAvailable"] = 92] = "NoMobileDeviceAvailable";
    EResult[EResult["NoMobileDevice"] = 92] = "NoMobileDevice";
    EResult[EResult["TimeIsOutOfSync"] = 93] = "TimeIsOutOfSync";
    EResult[EResult["TimeNotSynced"] = 93] = "TimeNotSynced";
    EResult[EResult["SMSCodeFailed"] = 94] = "SMSCodeFailed";
    EResult[EResult["TooManyAccountsAccessThisResource"] = 95] = "TooManyAccountsAccessThisResource";
    EResult[EResult["AccountLimitExceeded"] = 95] = "AccountLimitExceeded";
    EResult[EResult["AccountActivityLimitExceeded"] = 96] = "AccountActivityLimitExceeded";
    EResult[EResult["PhoneActivityLimitExceeded"] = 97] = "PhoneActivityLimitExceeded";
    EResult[EResult["RefundToWallet"] = 98] = "RefundToWallet";
    EResult[EResult["EmailSendFailure"] = 99] = "EmailSendFailure";
    EResult[EResult["NotSettled"] = 100] = "NotSettled";
    EResult[EResult["NeedCaptcha"] = 101] = "NeedCaptcha";
    EResult[EResult["GSLTDenied"] = 102] = "GSLTDenied";
    EResult[EResult["GSOwnerDenied"] = 103] = "GSOwnerDenied";
    EResult[EResult["InvalidItemType"] = 104] = "InvalidItemType";
    EResult[EResult["IPBanned"] = 105] = "IPBanned";
    EResult[EResult["GSLTExpired"] = 106] = "GSLTExpired";
    EResult[EResult["InsufficientFunds"] = 107] = "InsufficientFunds";
    EResult[EResult["TooManyPending"] = 108] = "TooManyPending";
    EResult[EResult["NoSiteLicensesFound"] = 109] = "NoSiteLicensesFound";
    EResult[EResult["WGNetworkSendExceeded"] = 110] = "WGNetworkSendExceeded";
    EResult[EResult["AccountNotFriends"] = 111] = "AccountNotFriends";
    EResult[EResult["LimitedUserAccount"] = 112] = "LimitedUserAccount";
    EResult[EResult["CantRemoveItem"] = 113] = "CantRemoveItem";
    EResult[EResult["AccountHasBeenDeleted"] = 114] = "AccountHasBeenDeleted";
    EResult[EResult["AccountDeleted"] = 114] = "AccountDeleted";
    EResult[EResult["AccountHasAnExistingUserCancelledLicense"] = 115] = "AccountHasAnExistingUserCancelledLicense";
    EResult[EResult["ExistingUserCancelledLicense"] = 115] = "ExistingUserCancelledLicense";
    EResult[EResult["DeniedDueToCommunityCooldown"] = 116] = "DeniedDueToCommunityCooldown";
    EResult[EResult["CommunityCooldown"] = 116] = "CommunityCooldown";
    EResult[EResult["NoLauncherSpecified"] = 117] = "NoLauncherSpecified";
    EResult[EResult["MustAgreeToSSA"] = 118] = "MustAgreeToSSA";
    EResult[EResult["ClientNoLongerSupported"] = 119] = "ClientNoLongerSupported";
    EResult[EResult["LauncherMigrated"] = 119] = "LauncherMigrated";
    EResult[EResult["CurrentSteamRealmDoesNotMatch"] = 120] = "CurrentSteamRealmDoesNotMatch";
    EResult[EResult["SteamRealmMismatch"] = 120] = "SteamRealmMismatch";
    EResult[EResult["InvalidSignature"] = 121] = "InvalidSignature";
    EResult[EResult["ParseFailure"] = 122] = "ParseFailure";
    EResult[EResult["NoVerifiedPhone"] = 123] = "NoVerifiedPhone";
    EResult[EResult["InsufficientBattery"] = 124] = "InsufficientBattery";
    EResult[EResult["ChargerRequired"] = 125] = "ChargerRequired";
    EResult[EResult["CachedCredentialInvalid"] = 126] = "CachedCredentialInvalid";
    EResult[EResult["PhoneNumberIsVOIP"] = 127] = "PhoneNumberIsVOIP";
})(EResult || (EResult = {}));
exports.default = EResult;
