// websocket13/dist/enums/StatusCode.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var StatusCode;
(function (StatusCode) {
    StatusCode[StatusCode["NormalClosure"] = 1000] = "NormalClosure";
    StatusCode[StatusCode["EndpointGoingAway"] = 1001] = "EndpointGoingAway";
    StatusCode[StatusCode["ProtocolError"] = 1002] = "ProtocolError";
    StatusCode[StatusCode["UnacceptableDataType"] = 1003] = "UnacceptableDataType";
    //Reserved1 = 1004,           /** Reserved. Do not use. */
    StatusCode[StatusCode["NoStatusCode"] = 1005] = "NoStatusCode";
    StatusCode[StatusCode["AbnormalTermination"] = 1006] = "AbnormalTermination";
    StatusCode[StatusCode["InconsistentData"] = 1007] = "InconsistentData";
    StatusCode[StatusCode["PolicyViolation"] = 1008] = "PolicyViolation";
    StatusCode[StatusCode["MessageTooBig"] = 1009] = "MessageTooBig";
    StatusCode[StatusCode["MissingExtension"] = 1010] = "MissingExtension";
    StatusCode[StatusCode["UnexpectedCondition"] = 1011] = "UnexpectedCondition";
    StatusCode[StatusCode["ServiceRestart"] = 1012] = "ServiceRestart";
    StatusCode[StatusCode["TryAgainLater"] = 1013] = "TryAgainLater";
    //Reserved2 = 1014,           /** Reserved. Do not use. */
    StatusCode[StatusCode["TLSFailed"] = 1015] = "TLSFailed"; /** MUST NOT be sent over the wire. Used internally when TLS handshake fails. */
})(StatusCode || (StatusCode = {}));
exports.default = StatusCode;
