// @doctormckay/stdlib/lib/http/client/CookieJar.js
"use strict";
var __classPrivateFieldSet = (this && this.__classPrivateFieldSet) || function (receiver, state, value, kind, f) {
    if (kind === "m") throw new TypeError("Private method is not writable");
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a setter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot write private member to an object whose class did not declare it");
    return (kind === "a" ? f.call(receiver, value) : f ? f.value = value : state.set(receiver, value)), value;
};
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
var _CookieJar_cookies;
Object.defineProperty(exports, "__esModule", { value: true });
const Cookie_1 = __importDefault(require("./Cookie"));
class CookieJar {
    constructor() {
        _CookieJar_cookies.set(this, void 0);
        __classPrivateFieldSet(this, _CookieJar_cookies, [], "f");
    }
    get cookies() {
        return __classPrivateFieldGet(this, _CookieJar_cookies, "f").slice(0);
    }
    add(cookie, domain) {
        if (typeof cookie == 'string') {
            cookie = Cookie_1.default.parse(cookie, domain);
        }
        let cookieObj = cookie;
        // If this matches an existing cookie by name and domain, delete that old one
        let matchingCookie = __classPrivateFieldGet(this, _CookieJar_cookies, "f").findIndex(c => c.name == cookieObj.name && c.domain == cookieObj.domain);
        if (matchingCookie != -1) {
            __classPrivateFieldGet(this, _CookieJar_cookies, "f").splice(matchingCookie, 1);
        }
        __classPrivateFieldGet(this, _CookieJar_cookies, "f").push(cookie);
    }
    remove(cookieName, domain) {
        let matchingCookie = __classPrivateFieldGet(this, _CookieJar_cookies, "f").findIndex(c => c.name == cookieName && c.domain == domain);
        if (matchingCookie != -1) {
            __classPrivateFieldGet(this, _CookieJar_cookies, "f").splice(matchingCookie, 1);
            return true;
        }
        return false;
    }
    getCookieHeaderForUrl(url) {
        let parsedUrl = new URL(url);
        return __classPrivateFieldGet(this, _CookieJar_cookies, "f").filter(cookie => cookie.shouldSendForRequest(parsedUrl.hostname, parsedUrl.pathname, parsedUrl.protocol == 'https:')).map(c => [c.name, c.content].join('=')).join('; ');
    }
    stringify() {
        return JSON.stringify(__classPrivateFieldGet(this, _CookieJar_cookies, "f").map(c => c.stringify()));
    }
    static parse(stringifiedJar) {
        let jar = new CookieJar();
        JSON.parse(stringifiedJar)
            .map(cookieString => Cookie_1.default.parse(cookieString, '__jarimport__'))
            .filter(v => v)
            .forEach(cookie => jar.add(cookie, cookie.domain));
        return jar;
    }
}
exports.default = CookieJar;
_CookieJar_cookies = new WeakMap();
Object.defineProperty(CookieJar.prototype, 'cookies', { enumerable: true });
